extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,319 @@
+
+    assert(this.element.hasAttribute('data-show-if-href'),
+        'amp-user-notification should have `data-show-if-href` attribute.');
+
+    /** @private @const {string} */
+    this.showIfHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-show-if-href'), this.element);
+
+    assert(this.element.hasAttribute('data-dismiss-href'),
+        'amp-user-notification should have `data-dismiss-href` attribute.');
+
+    /** @private @const {string} */
+    this.dismissHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-dismiss-href'), this.element);
+
+    this.options_ = {
+      credentials: this.element.getAttribute('credentials')
+    };
+
+    /** @private {?function(*)} */
+    this.resolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.dialogPromise_ = new Promise((resolve, reject) => {
+      this.resolve_ = resolve;
+    });
+
+    this.userNotification_.registerUserNotification(this.elementId_,
+        this.dialogPromise_);
+
+    this.registerAction('dismiss', this.dismiss.bind(this));
+  }
+
+  /** @override */
+  layoutCallback(visible) {
+    if (!this.activated_) {
+      this.activated_ = true;
+      return this.getAsyncCid_()
+          .then(this.fetchShowEndpoint_.bind(this))
+          .then(this.onFetchShowEndpointSuccess_.bind(this),
+              this.onRejection_.bind(this));
+    }
+    return Promise.resolve();
+  }
+
+  /**
+   * Executes a `GET` request to the href given on the `data-show-if-href`
+   * attribute.
+   * @param {string} ampUserId
+   * @return {!Promise}
+   * @private
+   */
+  fetchShowEndpoint_(ampUserId) {
+    this.ampUserId_ = ampUserId;
+    const getReq = {
+      /** @type {GetRequestMetadata} */
+      body: {
+        elementId: this.elementId_,
+        ampUserId: ampUserId
+      }
+    };
+
+    return xhrFor(this.win_).fetchJson(this.showIfHref_, getReq);
+  }
+
+  /**
+   * Creates an POST to the specified `data-dismiss-href` url.
+   * @private
+   * @return {!Promise}
+   */
+  postDismissEnpoint_() {
+    return xhrFor(this.win_).fetchJson(this.dismissHref_, {
+      method: 'POST',
+      body: {
+        elementId: this.elementId_,
+        ampUserId: this.ampUserId_
+      }
+    });
+  }
+
+  /**
+   * Success handler for `fetchShowEndpoint_`.
+   * @param {!GetResponseMetadata}
+   * @return {!Promise|!GetResponseMetadata}
+   * @private
+   */
+  onFetchShowEndpointSuccess_(data) {
+    assert(typeof data.showNotification == 'boolean', '`showNotification` ' +
+        'should be a boolean. Got "%s" which is of type %s.',
+        data.showNotification, typeof data.showNotification);
+
+    if (data.showNotification) {
+      this.show();
+      return data;
+    }
+    // if no notification is shown, assume consent.
+    return this.resolve_();
+  }
+
+  /**
+   * Generic rejection handler for amp-user-notification.
+   * @private
+   */
+  onRejection_(msg) {
+    console/*OK*/.error('Rejection ocurred on `amp-user-notification` ' +
```

Comment:
will reuse `log` instead of `console.error` to change it from `pub-dev` error to `amp-dev` error.